WreckEngine/UI.cpp

```diff
@@ -0,0 +1,402 @@
+#include "UI.h"
+#include "gl_structs.h"
+#include "GLError.h"
+#include "External.h"
+#if defined(_WIN32) || defined(_WIN64)
+#define GLFW_EXPOSE_NATIVE_WIN32
+#include <GLFW/glfw3native.h>
+#endif
+
+#if defined(offsetof)
+#define OFFSETOF(TYPE, ELEMENT) offsetof(TYPE, ELEMENT)
+#else
+#define OFFSETOF(TYPE, ELEMENT) ((size_t)&(((TYPE *)0)->ELEMENT))
+#endif
+
+namespace UI
+{
+    // NOTE - Shaders and globals are copied from ImGui's GLFW example
+
+    static const char* const ImGuiVertexShader =
+        "#version 330\n"
+        "uniform mat4 ProjMtx;\n"
+        "in vec2 Position;\n"
+        "in vec2 UV;\n"
+        "in vec4 Color;\n"
+        "out vec2 Frag_UV;\n"
+        "out vec4 Frag_Color;\n"
+        "void main()\n"
+        "{\n"
+        "   Frag_UV = UV;\n"
+        "   Frag_Color = Color;\n"
+        "   gl_Position = ProjMtx * vec4(Position.xy,0,1);\n"
+        "}\n";
+
+    static const char* const ImGuiFragmentShader =
+        "#version 330\n"
+        "uniform sampler2D Texture;\n"
+        "in vec2 Frag_UV;\n"
+        "in vec4 Frag_Color;\n"
+        "out vec4 Out_Color;\n"
+        "void main()\n"
+        "{\n"
+        "   Out_Color = Frag_Color * texture( Texture, Frag_UV.st);\n"
+        "}\n";
+
+    static double   g_Time = 0.0;
+    static bool     g_MousePressed[3] = { false, false, false };
+    static float    g_MouseWheel = 0.0f;
+    static GLuint   g_FontTexture = 0;
+    static int      g_ShaderHandle = 0, g_VertHandle = 0, g_FragHandle = 0;
+    static int      g_AttribLocationTex = 0, g_AttribLocationProjMtx = 0;
+    static int      g_AttribLocationPosition = 0, g_AttribLocationUV = 0, g_AttribLocationColor = 0;
+    static GLuint   g_VboHandle = 0, g_VaoHandle = 0, g_ElementsHandle = 0;
+
+    /**
+     * \brief The implementation for retrieving the clipboard's text for ImGui.
+     *
+     * \param userData ImGui user data.
+     * \return The clipboard's contents.
+     */
+    static const char* ImGuiGetClipboardImpl(void* userData)
+    {
+        return glfwGetClipboardString(Window::window);
+    }
+
+    /**
+     * \brief The implementation for setting the clipboard's text for ImGui.
+     *
+     * \param userData ImGui user data.
+     * \param contents The clipboard's new contents.
+     */
+    static void ImGuiSetClipboardImpl(void* userData, const char* contents)
+    {
+        glfwSetClipboardString(Window::window, contents);
+    }
+
+    /**
+     * \brief Renders the given IMGUI draw data.
+     *
+     * \param dd The draw data.
+     */
+    void ImGuiRenderImpl(ImDrawData* dd)
+    {
+        // NOTE - This function is basically the same as ImGui's GLFW example
+
+        GLsavestate glStateHelper;
+
+        // Avoid rendering when minimized, scale coordinates for retina displays (screen coordinates != framebuffer coordinates)
+        ImGuiIO& io = ImGui::GetIO();
```

Comment:
I mean you're gonna have to do that at some point... or scroll/look up to see what the type is. hovering is usually the easier option. But whatever man, it's up to you in this case, just try to use it for internal types; it saved me a lot of time as I was refactoring my own stuff.